cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -134,6 +142,55 @@
     this.tasks = new LinkedList<>();
     this.metricsCollectionService = metricsCollectionService;
     this.programCompletionNotifiers = Collections.emptySet();
+    this.cConf = cConf;
+    this.txRunner = transactionRunner;
+    this.store = store;
+  }
+
+  @Override
+  protected void doStartUp() throws Exception {
+    super.doStartUp();
+    AtomicBoolean completed = new AtomicBoolean();
+    AtomicReference<AppMetadataStore.Cursor> cursorRef = new AtomicReference<>(AppMetadataStore.Cursor.EMPTY);
+    int limit = cConf.getInt(Constants.RuntimeMonitor.INIT_BATCH_SIZE);
+    long startTs = System.currentTimeMillis();
+
+    try {
+      while (!completed.get()) {
+        TransactionRunners.run(txRunner, context -> {
```

Comment:
nit: do you need a retry strategy for this whole transaction as it involves external dependency sql.

Reply:
Good catch. Added retry strategy using `system.runtime.monitor.*` configs